### PR TITLE
Add ESLint rule to prevent usage of the verb 'toggle' in translatable strings

### LINKED ISSUE #66372: Add linting rules to prevent the usage of the verb 'toggle' in all translatable strings
### Description

See 
https://core.trac.wordpress.org/ticket/34753
https://github.com/WordPress/gutenberg/discussions/42492
https://github.com/WordPress/gutenberg/issues/66369
https://github.com/WordPress/gutenberg/pull/66371

It would be nice to add ESLint and PHP linting rules to prevent future usages of the verb 'toggle' inside translatable strings.

### Step-by-step reproduction instructions

N/A

### Screenshots, screen recording, code snippet

_No response_

### Environment info

_No response_

### Please confirm that you have searched existing issues in the repo.

- [x] Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

- [x] Yes

### Please confirm which theme type you used for testing.

- [ ] Block
- [ ] Classic
- [ ] Hybrid (e.g. classic with theme.json)
- [ ] Not sure

### PR DESCRIPTION
**Pull request by im3dabasia**
Closes : #66372 

## What?
As described in the above issue, the Gutenberg repository moved away from using the verb "toggle" because it was causing issues with translations. A new issue was proposed (#66372) to add an ESLint rule to prevent contributors from using this word in translation-ready functions like __().

## Why?
The usage of the verb "toggle" was causing incorrect translations in the core, as it does not translate well in many languages.

## How?
An ESLint rule was added to prevent the usage of the verb "toggle."

## Screenshots or screencast 
image

https://github.com/user-attachments/assets/8e502043-75a7-4717-b4cc-d91ef99c5b6b

### CODE REVIEW

**Comment by im3dabasia** [comment]
Hey @afercia,
Could you please review this PR when you have a moment?

**Comment by Mamaduka** [comment]
@swissspidy, I think you've got the most experience with custom i18n ESLint rules. Do you mind having a look at this?

**Review by swissspidy — CHANGES_REQUESTED** [review]
IMHO such a rule is too niche/specific for this specific string. If anything, it should use a user-configurable blocklist. It's also very specific for WordPress/Gutenberg, so I don't think `eslint-plugin` is a good fit for it.

I would just use `no-restricted-syntax` for this. There's a lot of precedent for this in `eslintrc.js`:

https://github.com/WordPress/gutenberg/blob/2ee383c032df762b77b5bd6c480582e0772a0db0/.eslintrc.js#L86-L140

The following rule works just fine:

```
	{
		selector: 'Literal[value=/^toggle/i]',
		message: "Avoid using the verb 'Toggle' in translatable strings",
	},
```

**Comment by im3dabasia** [comment]
Hi @swissspidy,

I used `Literal[value=/^toggle\\b/i]` to ensure that the rule only matches the word "toggle" when it appears standalone, excluding cases like "ToggleControl."

This approach helps avoid false positives in translatable strings.

Please take a look at my PR and let me know your feedback.

**Comment by swissspidy** [comment]
Looks much better! Now we just need to fix the 20 newly reported errors :)

**Comment by im3dabasia** [comment]
@swissspidy,

Upon reviewing the errors flagged by the current rule, it appears that the rule is raising errors for the word "toggle" whenever it appears in any string literal, including cases where it is not inside a translation function (such as in Examples 11, 12, 13, etc.).

However, our initial goal was to only prevent the use of the word "toggle" in translation-ready functions like __() and similar.

Could you please advise on how to adjust the rule to restrict the error only to those specific translation contexts, and prevent it from flagging usage of "toggle" in other non-translation-related string literals?

Thank you in advance for your guidance!

1. https://github.com/WordPress/gutenberg/blob/287d6534f59ea95de5499569f49996e7fa320a7c/packages/block-editor/src/components/responsive-block-control/index.js#L59-L61
2. https://github.com/WordPress/gutenberg/blob/287d6534f59ea95de5499569f49996e7fa320a7c/packages/block-library/src/navigation/edit/test/navigation-menu-selector.js#L73-L90
3. https://github.com/WordPress/gutenberg/blob/287d6534f59ea95de5499569f49996e7fa320a7c/packages/block-library/src/query-pagination/query-pagination-label-control.js#L12-L14
4. https://github.com/WordPress/gutenberg/blob/84fc1975e2a4fedc4d0b1cc09bd99e7a023d0d31/packages/block-library/src/table-of-contents/edit.js#L119-L127
5. https://github.com/WordPress/gutenberg/blob/84fc1975e2a4fedc4d0b1cc09bd99e7a023d0d31/packages/components/src/composite/test/index.tsx#L175-L177
6. https://github.com/WordPress/gutenberg/blob/84fc1975e2a4fedc4d0b1cc09bd99e7a023d0d31/packages/components/src/date-time/date/test/use-lilius.ts#L319-L320
7. https://github.com/WordPress/gutenberg/blob/84fc1975e2a4fedc4d0b1cc09bd99e7a023d0d31/packages/components/src/dropdown/test/index.tsx#L72-L73
8. https://github.com/WordPress/gutenberg/blob/84fc1975e2a4fedc4d0b1cc09bd99e7a023d0d31/packages/components/src/modal/test/index.tsx#L329-L331
9. https://github.com/WordPress/gutenberg/blob/84fc1975e2a4fedc4d0b1cc09bd99e7a023d0d31/packages/components/src/modal/test/index.tsx#L343-L345
10. https://github.com/WordPress/gutenberg/blob/84fc1975e2a4fedc4d0b1cc09bd99e7a023d0d31/packages/components/src/modal/test/index.tsx#L385-L387
11. https://github.com/WordPress/gutenberg/blob/84fc1975e2a4fedc4d0b1cc09bd99e7a023d0d31/packages/components/src/toggle-group-control/toggle-group-control-option-base/component.tsx#L56-L59
12. https://github.com/WordPress/gutenberg/blob/84fc1975e2a4fedc4d0b1cc09bd99e7a023d0d31/packages/components/src/toggle-group-control/toggle-group-control/as-button-group.tsx#L38-L41
13. https://github.com/WordPress/gutenberg/blob/84fc1975e2a4fedc4d0b1cc09bd99e7a023d0d31/packages/components/src/toggle-group-control/toggle-group-control/as-radio-group.tsx#L45-L48
14. https://github.com/WordPress/gutenberg/blob/84fc1975e2a4fedc4d0b1cc09bd99e7a023d0d31/packages/data/src/components/use-select/test/index.js#L129-L139
15. https://github.com/WordPress/gutenberg/blob/84fc1975e2a4fedc4d0b1cc09bd99e7a023d0d31/packages/edit-post/src/components/preferences-modal/enable-custom-fields.js#L19-L22
16. https://github.com/WordPress/gutenberg/blob/84fc1975e2a4fedc4d0b1cc09bd99e7a023d0d31/packages/edit-post/src/components/preferences-modal/test/enable-custom-fields.js#L81-L83
17. https://github.com/WordPress/gutenberg/blob/84fc1975e2a4fedc4d0b1cc09bd99e7a023d0d31/packages/editor/src/components/post-publish-button/post-publish-button-or-toggle.js#L13-L14
18. https://github.com/WordPress/gutenberg/blob/84fc1975e2a4fedc4d0b1cc09bd99e7a023d0d31/packages/editor/src/components/post-publish-button/test/post-publish-button-or-toggle.js#L58-L67
19. https://github.com/WordPress/gutenberg/blob/84fc1975e2a4fedc4d0b1cc09bd99e7a023d0d31/packages/token-list/src/test/index.ts#L243-L252

**Comment by swissspidy** [comment]
So how about `CallExpression[callee.name=/^(__|_x|_n|_nx)$/] > Literal[value=/^toggle\b/i]`

**Comment by im3dabasia** [comment]
Thank you very much, @swissspidy. I really appreciate it.

After adopting the new approach, the errors have been reduced to three, which are string literals in the __() functions. 

I believe these might have been missed previously.


1. `'Toggle between using the same value for all screen sizes or using a unique value per screen size.'`
2. `'Toggle off to hide the label text, e.g. "Next Page".'`
3. `'Toggle to only include headings from the current page (if the post is paginated).' `

<details>
  <summary>Click to view code</summary>

 https://github.com/WordPress/gutenberg/blob/287d6534f59ea95de5499569f49996e7fa320a7c/packages/block-editor/src/components/responsive-block-control/index.js#L59-L61

 https://github.com/WordPress/gutenberg/blob/287d6534f59ea95de5499569f49996e7fa320a7c/packages/block-library/src/query-pagination/query-pagination-label-control.js#L12-L14

 https://github.com/WordPress/gutenberg/blob/84fc1975e2a4fedc4d0b1cc09bd99e7a023d0d31/packages/block-library/src/table-of-contents/edit.js#L119-L127
</details>

**Comment by swissspidy** [comment]
So can we fix them? :-)

**Comment by Mamaduka** [comment]
The Query Pagination text has a separate issue - #51335. I'm working on a fix - #68105.

**Comment by im3dabasia** [comment]
Could you please suggest which of the following revised strings would be the most appropriate?

1. **Original:**  
   `'Toggle between using the same value for all screen sizes or using a unique value per screen size.'`

   **Revised Alternatives:**
   - `'Choose whether to use the same value for all screen sizes or a unique value for each screen size.'`
   - `'Select whether to apply the same value across all screen sizes or a different value for each size.'`

2. **Original:**  (@Mamaduka is working on this)
   `'Toggle off to hide the label text, e.g. "Next Page".'`

   **Revised Alternatives:**
   - `'Disable to hide the label text, such as "Next Page".'`
   - `'Turn off to hide the label text (e.g., "Next Page").'` 

3. **Original:**  
   `'Toggle to only include headings from the current page (if the post is paginated).'`

   **Revised Alternatives:**
   - `'Select to include only headings from the current page (if the post is paginated). '`
   - `'Enable to include only headings from the current page (if the post is paginated). '`

**Comment by swissspidy** [comment]
I'd say:

* `'Choose whether to use the same value for all screen sizes or a unique value for each screen size.'`
* `Only include headings from the current page (if the post is paginated).`

File: packages/block-library/src/table-of-contents/edit.js [hunk 1]
```diff
@@ -122,7 +122,7 @@ export default function TableOfContentsEdit( {
 									'Only including headings from the current page (if the post is paginated).'
 							  )
 							: __(
-									'Toggle to only include headings from the current page (if the post is paginated).'
+									'Only include headings from the current page (if the post is paginated).'
```
Comment (swissspidy) [inline packages/block-library/src/table-of-contents/edit.js]: Apologies, now with the correct context I see that my suggestion didn't make sense at all.

`Include headings from all pages (if the post is paginated).` makes more sense.

Comment (im3dabasia) [reply packages/block-library/src/table-of-contents/edit.js]: Yeah, that’s correct. It had been wrong before our change. Nice catch!

Comment (im3dabasia) [reply packages/block-library/src/table-of-contents/edit.js]: @swissspidy , I have made the suggested changes. Please have a look.

**Comment by Mamaduka** [comment]
@im3dabasia, the Query Pagination fix has been merged. Rebasing will resolve the failing Static Analysis check.

**Review by swissspidy — APPROVED** [review]

**Comment by im3dabasia** [comment]
> @im3dabasia, the Query Pagination fix has been merged. Rebasing will resolve the failing Static Analysis check.

Thank you for the update. I will go ahead and rebase my branch right away

**Comment by im3dabasia** [comment]
Hey @Mamaduka and @swissspidy,

I wanted to bring up an issue I noticed while working on [this PR](https://github.com/WordPress/gutenberg/pull/68894). The ESLint rule that prevents the use of the verb "toggle" has a flaw in its regex implementation.

Currently, the rule only flags instances where "toggle" appears at the beginning of a string (e.g., "toggle here to see the results"). However, it does not trigger a warning when "toggle" appears elsewhere in the string (e.g., "adding some words. toggle now").

### Before Screencast

https://github.com/user-attachments/assets/31b66721-567d-4bdd-ac31-4a1bb9d322af

### After Screencast

After updating the regex to `Literal[value=/\\btoggle\\b/i]`, it now works as expected.

https://github.com/user-attachments/assets/9d0d7127-6217-4f0d-bdc2-19913d292079


Let me know your thoughts! If needed, I’d be happy to open a follow-up PR to address this. 🚀

**Comment by swissspidy** [comment]
@im3dabasia Yes please, a follow-up PR would be great. It's just one character (`^`) that needs to be removed from the regex

**Comment by im3dabasia** [comment]
>Yes please, a follow-up PR would be great. It's just one character (`^`) that needs to be removed from the regex

@swissspidy , Including `\\b` would help us too, because we don't want to raise false positive alerts. Having `\\b` prevents words like "toggled" or "toggles" from being flagged unnecessarily. I would agree that `\\b` is not required at the front of "toggle" but in my opinion, it makes sense to have it at the back. What do you think?


image

**Comment by swissspidy** [comment]
Sure we can try that. Best to open a PR and then see the errors & iterate from there.

**Pull request merged**